src/malli/core.cljc

```diff
@@ -275,6 +275,47 @@
           (-properties [_] properties)
           (-form [_] form))))))
 
+(defn- -map-of-schema []
+  ^{:type ::into-schema}
+  (reify IntoSchema
+    (-name [_] :map-of)
+    (-into-schema [_ properties childs opts]
+      (when-not (and (seq childs) (= 2 (count childs)))
+        (fail! ::invalid-map-of))
+      (let [[key-schema value-schema :as schemas] (mapv #(schema % opts) childs)
+            key-valid? (-validator key-schema)
+            value-valid? (-validator value-schema)
+            validate (fn [m]
+                       (reduce-kv
+                         (fn [_ key value]
+                           (or (and (key-valid? key) (value-valid? value)) (reduced false)))
+                         true m))]
+        ^{:type ::schema}
+        (reify Schema
+          (-validator [_] (fn [m] (and (map? m) (validate m))))
+          (-explainer [this path]
+            (let [distance (if (seq properties) 2 1)
+                  key-explainer (partial -explainer key-schema)
+                  value-explainer (partial -explainer value-schema)]
+              (fn explain [m in acc]
+                (if-not (map? m)
+                  (conj acc {:path   path
+                             :in     in
+                             :schema this
+                             :value  m
+                             :type   ::invalid-type})
+                  (reduce-kv
+                    (fn [acc key value]
+                      (let [in (conj in key)
+                            key-explainer' (key-explainer (conj path distance))
```

Comment:
These actually only use data which is available already at line 296->, and outside of the returned function. Lines 298 & 299 can create the concrete explainers already.